src/core/geometry/qgsnurbscurve.cpp

```diff
@@ -0,0 +1,1622 @@
+#include "qgswkbptr.h"
+#include "qgswkbtypes.h"
+
+#include <QPainterPath>
+
+using namespace nlohmann;
+
+QgsNurbsCurve::QgsNurbsCurve()
+{
+  mWkbType = Qgis::WkbType::NurbsCurve;
+}
+
+QgsNurbsCurve::QgsNurbsCurve( const QVector<QgsPoint> &ctrlPoints, int degree, const QVector<double> &knots, const QVector<double> &weights, bool closed )
+  : mControlPoints( ctrlPoints )
+  , mKnots( knots )
+  , mWeights( weights )
+  , mDegree( degree )
+  , mClosed( closed )
+{
+  mWkbType = Qgis::WkbType::NurbsCurve;
+
+  // Update WKB type based on coordinate dimensions
+  if ( !mControlPoints.isEmpty() )
+  {
+    const QgsPoint &firstPoint = mControlPoints.first();
+    if ( firstPoint.is3D() )
+      mWkbType = QgsWkbTypes::addZ( mWkbType );
+    if ( firstPoint.isMeasure() )
+      mWkbType = QgsWkbTypes::addM( mWkbType );
+  }
+}
+
+QgsCurve *QgsNurbsCurve::clone() const
+{
+  return new QgsNurbsCurve( *this );
+}
+
+/**
+ * \brief Find the knot span index for parameter u using binary search.
+ *
+ * This is Algorithm A2.1 from "The NURBS Book" (Piegl & Tiller).
+ * Returns the index i such that knots[i] <= u < knots[i+1].
+ * For a NURBS curve with n control points and degree p, valid spans
+ * are in range [p, n-1].
+ *
+ * \param degree polynomial degree of the NURBS curve
+ * \param u parameter value to locate
+ * \param knots knot vector
+ * \param nPoints number of control points
+ * \returns knot span index
+ */
+static int findKnotSpan( const int degree, const double u, const QVector<double> &knots, const int nPoints )
+{
+  // Special case: u at or beyond end of parameter range
+  if ( u >= knots[nPoints] )
+    return nPoints - 1;
+
+  // Special case: u at or before start of parameter range
+  if ( u <= knots[degree] )
+    return degree;
+
+  // Binary search for the knot span
+  int low = degree;
+  int high = nPoints;
+  int mid = ( low + high ) / 2;
+
+  while ( u < knots[mid] || u >= knots[mid + 1] )
+  {
+    if ( u < knots[mid] )
+      high = mid;
+    else
+      low = mid;
+    mid = ( low + high ) / 2;
+  }
+
+  return mid;
+}
+
+// Evaluates the NURBS curve at parameter t using De Boor's algorithm.
+// Implements Algorithm A4.1 from "The NURBS Book" (Piegl & Tiller).
+QgsPoint QgsNurbsCurve::evaluate( double t ) const
+{
+  const int n = mControlPoints.size();
+  if ( n == 0 )
+  {
+    return QgsPoint();
+  }
+
+  // Clamp parameter t to valid range [0,1]
+  if ( t <= 0.0 )
+    return mControlPoints.first();
+  if ( t >= 1.0 )
+    return mControlPoints.last();
+
+  const bool hasZ = !mControlPoints.isEmpty() && mControlPoints.first().is3D();
+  const bool hasM = !mControlPoints.isEmpty() && mControlPoints.first().isMeasure();
+
+  // Remap parameter from [0,1] to knot vector range [knots[degree], knots[n]]
+  const double u = mKnots[mDegree] + t * ( mKnots[n] - mKnots[mDegree] );
+
+  // Find the knot span containing parameter u (Algorithm A2.1)
+  const int span = findKnotSpan( mDegree, u, mKnots, n );
```

Comment:
I believe findKnotSpan() might crash if isValid() is false, because it assumes that the size of mKnots is consistant with n

And similar for access below. So it seems to be me this method should check and fail if isValid() is false. For performance reaons, I guess you would want isValid() to set a mValid flag, and invalidate it in clearCache()